README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> RabbitMQ specific components for Dart

This library is a part of [Pip.Services](https://github.com/pip-services/pip-services) project.
The RabbitMQ module contains a set of components for working with the message queue in RabbitMQ through the AMQP protocol.

The module contains the following packages:
- **Build** - Factory for constructing module components
- **Queue** - Message Queuing components that implement the standard [Messaging](https://github.com/pip-services3-dotnet/pip-services3-messaging-dotnet) module interface

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pub.dev/documentation/pip_services3_rabbitmq/latest/pip_services3_rabbitmq/pip_services3_rabbitmq-library.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)


## Use

Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  pip_services3_rabbitmq: version
```

Now you can install package from the command line:
```bash
pub get
```

## Develop

For development you shall install the following prerequisites:
* Dart SDK 2
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pub get
```

Run automated tests:
```bash
pub run test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The Dart version of Pip.Services is created and maintained by:
- **Levichev Dmitry**
- **Sergey Seroukhov**

The documentation is written by:
- **Levichev Dmitry**

Many thanks to contibutors, who put their time and talant into making this project better:
- **Andrew Harrington**, Kyrio

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> RabbitMQ specific components for Changelog

## 1.0.0

- Initial version, created by Sergey Seroukhov and Dmitry Levichev

## 1.1.0

- Migrated to null safety

test/queues/MessageQueueFixture.dart
import 'dart:async';

import 'package:test/test.dart';
import 'package:pip_services3_messaging/pip_services3_messaging.dart';

class TestMessageReciver implements IMessageReceiver {
  late MessageEnvelope message;

  @override
  Future receiveMessage(MessageEnvelope envelope, IMessageQueue queue) async {
    message = envelope;
  }
}

class MessageQueueFixture {
  IMessageQueue _queue;

  MessageQueueFixture(IMessageQueue queue) : _queue = queue;

  Future testSendReceiveMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;

    await _queue.send(null, envelope1);

    var count = await _queue.readMessageCount();
    expect(count > 0, isTrue);

    var result = await _queue.receive(null, 10000);
    envelope2 = result!;
    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);
  }

  Future testReceiveSendMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;

    await Future.delayed(Duration(milliseconds: 500), () async {
      await _queue.send(null, envelope1);
    });

    var result = await _queue.receive(null, 10000);
    envelope2 = result!;

    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);
  }

  Future testReceiveCompleteMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;

    await _queue.send(null, envelope1);

    var count = await _queue.readMessageCount();
    expect(count > 0, isTrue);

    var result = await _queue.receive(null, 10000);
    envelope2 = result!;

    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);

    await _queue.complete(envelope2);
    expect(envelope2.getReference(), isNull);
  }

  Future testReceiveAbandonMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;

    await _queue.send(null, envelope1);

    var result = await _queue.receive(null, 10000);
    envelope2 = result!;

    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);

    await _queue.abandon(envelope2);

    result = await _queue.receive(null, 10000);
    envelope2 = result!;

    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);
  }

  Future testSendPeekMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;

    await _queue.send(null, envelope1);

    var result = await _queue.peek(null);
    envelope2 = result!;

    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);
  }

  Future testPeekNoMessage() async {
    var result = await _queue.peek(null);
    expect(result, isNull);
  }

  Future testMoveToDeadMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;

    await _queue.send(null, envelope1);

    var result = await _queue.receive(null, 10000);
    envelope2 = result!;

    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);

    await _queue.moveToDeadLetter(envelope2);
  }

  Future testOnMessage() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    MessageEnvelope envelope2;
    var reciver = TestMessageReciver();
    _queue.beginListen(null, reciver);

    await Future.delayed(Duration(milliseconds: 1000), () {});
    await _queue.send(null, envelope1);

    await Future.delayed(Duration(milliseconds: 1000), () {});

    envelope2 = reciver.message;
    expect(envelope2, isNotNull);
    expect(envelope1.message_type, envelope2.message_type);
    expect(envelope1.message, envelope2.message);
    expect(envelope1.correlation_id, envelope2.correlation_id);

    _queue.endListen(null);
  }

  Future testMessageCount() async {
    var envelope1 = MessageEnvelope('123', 'Test', 'Test message');
    await _queue.send('', envelope1);
    await Future.delayed(Duration(milliseconds: 500), () {});
    var count = await _queue.readMessageCount();
    expect(count >= 1, true);
  }
}


test/queues/RabbitMQMessageQueue_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_rabbitmq/pip_services3_rabbitmq.dart';
import './MessageQueueFixture.dart';

void main() {
  group('Test RabbitMQMessageQueue', () {
    late RabbitMQMessageQueue queue;
    late MessageQueueFixture fixture;

    setUpAll(() async {
      var rabbitmqHost = Platform.environment['RABBITMQ_HOST'] ?? 'localhost';
      var rabbitmqPort = Platform.environment['RABBITMQ_PORT'] ?? '5672';
      var rabbitmqExchange =
          Platform.environment['RABBITMQ_EXCHANGE'] ?? 'test';
      var rabbitmqQueue = Platform.environment['RABBITMQ_QUEUE'] ?? 'test';
      var rabbitmqUser = Platform.environment['RABBITMQ_USER'] ?? 'user';
      var rabbitmqPassword = Platform.environment['RABBITMQ_PASS'] ?? 'pass123';

      if (rabbitmqHost == null && rabbitmqPort == null) {
        return;
      }
      var queueConfig = ConfigParams.fromTuples([
        'exchange', rabbitmqExchange,
        'queue', rabbitmqQueue,
        'options.auto_create', true,
        //'connection.protocol', 'amqp',
        'connection.host', rabbitmqHost,
        'connection.port', rabbitmqPort,
        'credential.username', rabbitmqUser,
        'credential.password', rabbitmqPassword,
      ]);
      queue = RabbitMQMessageQueue('testQueue');
      queue.configure(queueConfig);
      fixture = MessageQueueFixture(queue);

      await queue.open(null);
    });

    tearDownAll(() async {
      await queue.close(null);
    });

    setUp(() async {
      await queue.clear(null);
    });

    // test('RabbitMQMessageQueue:Send Receive Message', () async {
    //   await fixture.testSendReceiveMessage();
    // });

    // test('RabbitMQMessageQueue:Receive Send Message', () async {
    //   await fixture.testReceiveSendMessage();
    // });

    // test('RabbitMQMessageQueue:Receive And Complete Message', () async {
    //   await fixture.testReceiveCompleteMessage();
    // });

    // test('RabbitMQMessageQueue:Receive And Abandon Message', () async {
    //   await fixture.testReceiveAbandonMessage();
    // });

    // test('RabbitMQMessageQueue:Send Peek Message', () async {
    //   await fixture.testSendPeekMessage();
    // });

    // test('RabbitMQMessageQueue:Peek No Message', () async {
    //   await fixture.testPeekNoMessage();
    // });

    test('RabbitMQMessageQueue:On Message', () async {
      await fixture.testOnMessage();
    });
  });
}


lib/src/connect/RabbitMQConnectionResolver.dart
import 'dart:async';

import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

/// RabbitMQConnectionResolver helper class that resolves RabbitMQ connection and credential parameters,
/// validates them and generates connection options.
///   Configuration parameters:
/// - [connection(s)]:
///   - [discovery_key]:               (optional) a key to retrieve the connection from IDiscovery
///   - [host]:                        host name or IP address
///   - [port]:                        port number
///   - [uri]:                         resource URI or connection string with all parameters in it
/// - [credential(s)]:
///   - [store_key]:                   (optional) a key to retrieve the credentials from ICredentialStore
///   - [username]:                    user name
///   - [password]:                    user password
///  References:
/// - *:discovery:*:*:1.0          (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connections
/// - *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials [ICredentialStore](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICredentialStore-class.html)

class RabbitMQConnectionResolver implements IReferenceable, IConfigurable {
  // The connections resolver.
  ConnectionResolver connectionResolver;
  //The credentials resolver.
  CredentialResolver credentialResolver;

  RabbitMQConnectionResolver()
      : connectionResolver = ConnectionResolver(),
        credentialResolver = CredentialResolver();

  /// Configure are configures component by passing configuration parameters.
  /// Parameters:
  /// - [config]  configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    connectionResolver.configure(config);
    credentialResolver.configure(config);
  }

  /// SetReferences are sets references to dependent components.
  /// Parameters:
  /// - [references]  references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    connectionResolver.setReferences(references);
    credentialResolver.setReferences(references);
  }

  void _validateConnection(
      String? correlationId, ConnectionParams? connection) {
    if (connection == null) {
      throw ConfigException(
          correlationId, 'NO_CONNECTION', 'RabbitMQ connection is not set');
    }

    var uri = connection.getUri();
    if (uri != null && uri.isNotEmpty) {
      return;
    }

    var protocol = connection.getAsNullableString('protocol');
    if (protocol == null || protocol.isEmpty) {
      //return cerr.NewConfigError(correlationId, 'NO_PROTOCOL', 'Connection protocol is not set')
      connection.setAsObject('protocol', 'amqp');
    }

    var host = connection.getHost();
    if (host == null || host.isEmpty) {
      throw ConfigException(
          correlationId, 'NO_HOST', 'Connection host is not set');
    }

    var port = connection.getPort();
    if (port == 0) {
      throw ConfigException(
          correlationId, 'NO_PORT', 'Connection port is not set');
    }
  }

  ConfigParams _composeOptions(
      ConnectionParams? connection, CredentialParams? credential) {
    // Define additional parameters parameters
    credential ??= CredentialParams();

    var options = connection!.override(credential);

    // Compose uri
    if (options.get('uri') == null) {
      var credential = '';
      if (options.get('username') != null) {
        credential = options.get('username')!;
      }
      if (options.get('password') != null) {
        credential += ':' + options.get('password')!;
      }
      var uri = '';
      if (credential.isEmpty) {
        uri = options.get('protocol')! + '://' + options.get('host')!;
      } else {
        uri = options.get('protocol')! +
            '://' +
            credential +
            '@' +
            options.get('host')!;
      }
      if (options.get('port') != null) {
        uri = uri + ':' + options.get('port')!;
      }
      options.setAsObject('uri', uri);
    }
    return options;
  }

  /// Resolves RabbitMQ connection options from connection and credential parameters.
  /// Parameters:
  ///  - [correlationId]   (optional) transaction id to trace execution through call chain.
  /// Retruns               Future that receives resolved options
  /// Throw error.
  Future<ConfigParams> resolve(String? correlationId) async {
    ConnectionParams? connection;
    CredentialParams? credential;
    var err;

    await Future.wait([
      () async {
        try {
          connection = await connectionResolver.resolve(correlationId);
          //Validate connections
          _validateConnection(correlationId, connection);
        } catch (ex) {
          err = ex;
        }
      }(),
      () async {
        try {
          credential = await credentialResolver.lookup(correlationId);
          // Credentials are not validated right now
        } catch (ex) {
          err = ex;
        }
      }()
    ]);

    if (err != null) {
      throw err;
    }
    return _composeOptions(connection, credential);
  }

  /// Compose method are composes RabbitMQ connection options from connection and credential parameters.
  /// Parameters:
  ///    - [correlationId]    (optional) transaction id to trace execution through call chain.
  ///    - [connection]    connection parameters
  ///    - [credential]     credential parameters
  /// Returns               Future that receives resolved options
  /// Throw error.
  Future<ConfigParams> compose(String? correlationId,
      ConnectionParams? connection, CredentialParams? credential) async {
    // Validate connections
    _validateConnection(correlationId, connection);
    return _composeOptions(connection, credential);
  }
}


lib/src/queues/RabbitMQMessageQueueFactory.dart
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';
import './RabbitMQMessageQueue.dart';

class RabbitMQMessageQueueFactory extends Factory {
  static final descriptor = Descriptor(
      'pip-services3-rabbitmq', 'factory', 'message-queue', 'rabbitmq', '1.0');
  static final MemoryQueueDescriptor = Descriptor(
      'pip-services3-rabbitmq', 'message-queue', 'rabbitmq', '*', '*');

  ConfigParams? config;

  RabbitMQMessageQueueFactory() : super() {
    register(MemoryQueueDescriptor, (locator) {
      var queue = RabbitMQMessageQueue(locator.getName());
      if (config != null) {
        queue.configure(config!);
      }
      return queue;
    });
  }

  void configure(ConfigParams config) {
    this.config = config;
  }
}


lib/src/queues/RabbitMQMessageQueue.dart
import 'dart:async';

import 'package:dart_amqp/dart_amqp.dart' as amqp;
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';
import 'package:pip_services3_messaging/pip_services3_messaging.dart';
import '../connect/RabbitMQConnectionResolver.dart';

///  Message queue that sends and receives messages via RabbitMQ message broker.
///  RabbitMQ is a popular light-weight protocol to communicate.
///  Configuration parameters:
///
///  [connection(s)]:
///  - [discovery_key]:               (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
///  - [host]:                        host name or IP address
///  - [port]:                        port number
///  - [uri]:                         resource URI or connection string with all parameters in it
/// - [credential(s)]:
///  - [store_key]:                   (optional) a key to retrieve the credentials from [ICredentialStore](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICredentialStore-class.html)
///  - [username]:                    user name
///  - [password]:                    user password
///
///  References:
///
///  - *:logger:*:*:1.0             (optional) [ILogger](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ILogger-class.html) components to pass log messages
///  - *:counters:*:*:1.0           (optional) [ICounters](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICounters-class.html) components to pass collected measurements
///  - *:discovery:*:*:1.0          (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connections
///  - *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials
///
///  var queue = RabbitMQMessageQueue('my_queue');
///  queue.configure(ConfigParams.fromTuples([
///  'topic', 'mytopic',
///  'connection.protocol', 'amqp'
///  'connection.host', 'localhost'
///  'connection.port', 5672 ]));
///  await queue.open('123');
///
///  await queue.send('123', MessageEnvelop('123', 'mymessage', 'ABC'));
///  await queue.receive('123', 0);
///  await queue.complete('123', message);

class RabbitMQMessageQueue extends MessageQueue {
  final _defaultCheckinterval = 1000;
  amqp.Client? _connection;
  amqp.Channel? _mqChanel;
  RabbitMQConnectionResolver _optionsResolver;
  String? _queueName;
  String? _exchangeName;
  amqp.Queue? _queue;
  amqp.Exchange? _exchange;
  amqp.ExchangeType _exchangeType = amqp.ExchangeType.FANOUT;
  amqp.Consumer? _consumer;
  String? _routingKey;
  bool _persistent = false;
  bool _exclusive = false;
  bool _autoCreate = false;
  bool _autoDelete = false;
  bool _noQueue = false;
  late int interval;

  /// Creates a new instance of the message _queue.
  /// - [name]  (optional) a queue name.
  /// - [config] (optional)
  /// - [mqChanel] (optional) RrabbitMQ chanel
  /// - [queue] (optional)  RrabbitMQ queue name
  RabbitMQMessageQueue(String name,
      {ConfigParams? config, amqp.Channel? mqChanel, String? queue})
      : _optionsResolver = RabbitMQConnectionResolver(),
        super() {
    capabilities = MessagingCapabilities(
        true, true, false, false, false, false, true, false, true);
    interval = _defaultCheckinterval;
    if (config != null) {
      configure(config);
    }
    _mqChanel = mqChanel;
    _queueName = queue;
  }

  ///  Configures component by passing configuration parameters.
  /// - [config] configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    super.configure(config);

    interval = config.getAsLongWithDefault('interval', _defaultCheckinterval);

    _queueName = config.getAsNullableString('queue') ?? _queueName;
    _exchangeName = config.getAsNullableString('exchange') ?? _exchangeName;

    _exchangeType = amqp.ExchangeType.valueOf(config.getAsStringWithDefault(
        'options.exchange_type', _exchangeType.toString()));
    _routingKey =
        config.getAsNullableString('options.routing_key') ?? _routingKey;
    _persistent =
        config.getAsBooleanWithDefault('options.persistent', _persistent);
    _exclusive =
        config.getAsBooleanWithDefault('options.exclusive', _exclusive);
    _autoCreate =
        config.getAsBooleanWithDefault('options.auto_create', _autoCreate);
    _autoDelete =
        config.getAsBooleanWithDefault('options.auto_delete', _autoDelete);
    _noQueue = config.getAsBooleanWithDefault('options.no_queue', _noQueue);
  }

  void _checkOpened(String? correlationId) {
    if (_mqChanel == null) {
      throw InvalidStateException(
          correlationId, 'NOT_OPENED', 'The _queue is not opened');
    }
  }

  ///  Checks if the component is opened.
  ///  Retruns true if the component has been opened and false otherwise.
  @override
  bool isOpen() {
    return _connection != null && _mqChanel != null;
  }

  ///  Opens the component with given _connection and credential parameters.
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  ///  - [connection] connection parameters
  ///  - [credential] credential parameters
  /// Return            Future that recive null if all ok
  /// Throws error
  @override
  Future openWithParams(String? correlationId, ConnectionParams? connection,
      CredentialParams? credential) async {
    var options =
        await _optionsResolver.compose(correlationId, connection, credential);

    if (_queueName == null && _exchangeName == null) {
      throw ConfigException(correlationId, 'NO_QUEUE',
          'Queue or exchange are not defined in connection parameters');
    }

    var settings = amqp.ConnectionSettings();
    var uri = Uri();
    var url = options.get('uri');
    uri = uri.resolve(url!);
    settings.host = uri.host;
    settings.port = uri.port;
    if (uri.userInfo != '') {
      var auth = amqp.PlainAuthenticator(
          options.get('username')!, options.get('password')!);
      settings.authProvider = auth;
    }

    _connection = amqp.Client(settings: settings);
    await _connection!.connect();

    _mqChanel = await _connection!.channel();

    // Automatically create queue, exchange and binding
    if (_autoCreate) {
      if (_exchangeName != null) {
        _exchange = await _mqChanel!
            .exchange(_exchangeName!, _exchangeType, durable: _persistent);
      }
      if (!_noQueue) {
        if (_queueName == null) {
          _queue = await _mqChanel!.queue('',
              durable: _persistent,
              autoDelete: true,
              exclusive: true,
              noWait: false);

          _queueName = _queue?.name;
        } else {
          _queue = await _mqChanel!.queue(_queueName!,
              durable: _persistent,
              exclusive: _exclusive,
              autoDelete: _autoDelete,
              noWait: false);
        }

        _queue =
            await _queue!.bind(_exchange!, _routingKey ?? '', noWait: false);
      }
    }
    return null;
  }

  /// Close method are closes component and frees used resources.
  ///  Parameters:
  ///   - [correlationId] (optional) transaction id to trace execution through call chain.
  /// Return            Future that recive null if all ok
  /// Throws error
  @override
  Future close(String? correlationId) async {
    if (_mqChanel != null) {
      await _mqChanel!.close();
    }

    if (_connection != null) {
      await _connection!.close();
    }
    _connection = null;
    _mqChanel = null;
    logger.trace(correlationId, 'Closed _queue %s', [_queue]);
  }

  /// ReadMessageCount method are reads the current number of messages in the _queue to be delivered.
  /// Returns         Future that contains count number of messages
  /// Throws error.
  @override
  Future<int> readMessageCount() async {
    try {
      _checkOpened('');
    } catch (err) {
      logger.error('', err as Exception,
          'RabbitMQMessageQueue:MessageCount: ' + err.toString());
      rethrow;
    }

    if (_queue == null) {
      return 0;
    }

    return _queue!.messageCount;
  }

  MessageEnvelope? _toMessage(amqp.AmqpMessage? envelope) {
    if (envelope == null) {
      return null;
    }

    var message = MessageEnvelope(envelope.properties?.corellationId,
        envelope.properties?.type, envelope.payloadAsString);
    message.message_id = envelope.properties?.messageId ?? '';
    message.sent_time = DateTime.now().toUtc();
    message.setReference(envelope);

    return message;
  }

  ///  Send method are sends a message into the _queue.
  ///  Parameters:
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  ///  - [message] a message envelop to be sent.
  /// Return            Future that recive null if all ok
  /// Throws error
  @override
  Future send(String? correlationId, MessageEnvelope message) async {
    _checkOpened(correlationId);

    var messageProperties = amqp.MessageProperties();
    messageProperties.contentType = 'text/plain';

    if (message.correlation_id != null) {
      messageProperties.corellationId = message.correlation_id;
    }
    if (message.message_id != null && message.message_id.isNotEmpty) {
      messageProperties.messageId = message.message_id;
    }
    messageProperties.persistent = _persistent;
    if (message.message_type != null) {
      messageProperties.type = message.message_type;
    }
    _queue!.publish(message.message!, properties: messageProperties);

    counters.incrementOne('_queue.' + name + '.sent_messages');
    logger.debug(
        message.correlation_id, 'Sent message %s via %s', [message, this]);
  }

  ///  Peeks a single incoming message from the queue without removing it.
  ///  If there are no messages available in the queue it returns null.
  ///  Important: This method are not supported in this release!
  ///  Parameters:
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  /// Return            Future that recive a message
  /// Throws error
  @override
  Future<MessageEnvelope> peek(String? correlationId) async {
    // _checkOpened(correlationId);

    // var envelope = _mqChanel.get(_queue, false);
    // var message = _toMessage(envelope);
    // if (message != null) {
    //   logger.trace(
    //       message.correlation_id, 'Peeked message %s on %s', [message, name]);
    // }
    // return message;

    throw Exception('Method "peek" are not supported!');
  }

  ///  PeekBatch method are peeks multiple incoming messages from the _queue without removing them.
  ///  If there are no messages available in the _queue it returns an empty list.
  ///  Parameters:
  ///   - [correlationId] (optional) transaction id to trace execution through call chain.
  ///   - [messageCount] a maximum number of messages to peek.
  /// Return            Future that recive  a list with messages
  /// Throws error
  @override
  Future<List<MessageEnvelope>> peekBatch(
      String? correlationId, int messageCount) async {
//  _checkOpened(correlationId);
// 	var messages = <MessageEnvelope>[];
// 	for (;messageCount > 0;) {
// 		var envelope = _mqChanel.get(_queue, false);
// 		// if getErr != null || !ok {
// 		// 	err = getErr
// 		// 	break
// 		// }
// 		var message = _toMessage(envelope);
// 		messages.add(message);
// 		messageCount--;
// 	}
// 	logger.trace(correlationId, 'Peeked %s messages on %s', [messages.length, name]);
// 	return messages;
    throw Exception('Method "peekBatch" are not supported!');
  }

  ///  Receive method are receives an incoming message and removes it from the _queue.
  ///  Parameters:
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  ///  - [waitTimeout] a timeout in milliseconds to wait for a message to come.
  /// Return            Future that recive a message
  /// Throws error
  @override
  Future<MessageEnvelope?> receive(
      String? correlationId, int waitTimeout) async {
    _checkOpened(correlationId);

    MessageEnvelope? receivedMessage;

    _consumer ??= await _queue!.consume();

    await Future(() async {
      _consumer!.listen((amqp.AmqpMessage message) {
        message.ack();
        receivedMessage = _toMessage(message);
      });

      var start = 0;
      while (waitTimeout - start > 0 && receivedMessage == null) {
        await Future.delayed(Duration(milliseconds: 100));
        start += 100;
      }
    });

    // await _consumer!.channel.;

    return receivedMessage;
  }

  ///  Renews a lock on a message that makes it invisible from other receivers in the _queue.
  ///  This method is usually used to extend the message processing time.
  ///  Important: This method is not supported by RabbitMQ.
  ///  Parameters:
  ///  - [message] a message to extend its lock.
  ///  - [lockTimeout] a locking timeout in milliseconds.
  /// Return            Future that recive a null if all ok
  /// Throws error
  @override
  Future renewLock(MessageEnvelope message, int lockTimeout) async {
    // Operation is not supported
    return null;
  }

  ///  Returnes message into the queue and makes it available for all subscribers to receive it again.
  ///  This method is usually used to return a message which could not be processed at the moment
  ///  to repeat the attempt.Messages that cause unrecoverable errors shall be removed permanently
  ///  or/and send to dead letter _queue.
  ///  Parameters:
  ///  - [message] a message to return.
  /// Return            Future that recive a null if all ok
  /// Throws error
  @override
  Future abandon(MessageEnvelope message) async {
    _checkOpened('');

    // Make the message immediately visible
    amqp.AmqpMessage envelope = message.getReference();
    if (envelope != null) {
      envelope.reject(true);
      message.setReference(null);
      logger.trace(message.correlation_id, 'Abandoned message %s at %s',
          [message, name]);
    }
    return null;
  }

  ///  Permanently removes a message from the _queue.
  ///  This method is usually used to remove the message after successful processing.
  ///  Parameters:
  ///  - [message] a message to remove.
  /// Return            Future that recive a null if all done
  /// Throws error
  @override
  Future complete(MessageEnvelope message) async {
    _checkOpened('');

    amqp.AmqpMessage envelope = message.getReference();
    if (envelope != null) {
      envelope.ack();
      message.setReference(null);
      logger.trace(message.correlation_id, 'Completed message %s at %s',
          [message, name]);
    }
  }

  ///  Permanently removes a message from the _queue and sends it to dead letter _queue.
  ///  Important: This method is not supported by RabbitMQ.
  ///  Parameters:
  ///  - [message] a message to be removed.
  /// Return            Future that recive a null when all done
  /// Throws error
  @override
  Future moveToDeadLetter(MessageEnvelope message) async {
    _checkOpened('');
    // Operation is not supported
  }

  ///  Listens for incoming messages and blocks the current thread until _queue is closed.
  /// Parameters:
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  ///  Returns            Future that recive null on compleate
  /// Throws error
  @override
  Future listen(String? correlationId, IMessageReceiver receiver) async {
    try {
      _checkOpened('');
    } catch (err) {
      logger.error(correlationId, err as Exception,
          'RabbitMQMessageQueue:Listen: Can\'t start listen ' + err.toString());
      rethrow;
    }

    logger.debug(correlationId, 'Started listening messages at %s', [name]);
    try {
      _consumer = await _queue!.consume();
    } catch (err) {
      logger.error(
          correlationId,
          err as Exception,
          'RabbitMQMessageQueue:Listen: Can\'t consume to _queue' +
              err.toString());
      rethrow;
    }

    _consumer!.listen((amqp.AmqpMessage msg) {
      var message = _toMessage(msg);
      counters.incrementOne('_queue.' + name + '.received_messages');
      logger.debug(message?.correlation_id, 'Received message %s via %s',
          [message, name]);
      try {
        receiver.receiveMessage(message!, this);
      } catch (err) {
        logger.error(
            message?.correlation_id,
            err as Exception,
            'Processing received message %s error in _queue %s',
            [message, name]);
      }
      msg.ack();
    });
  }

  ///  Ends listening for incoming messages.
  ///  When this method is call listen unblocks the thread and execution continues.
  ///  Parameters:
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  /// Return            Future that recive a null when all done
  /// Throws error
  @override
  Future endListen(String? correlationId) async {
    if (_consumer != null) {
      try {
        await _consumer!.cancel();
      } catch (ex) {
        logger.error(
            correlationId, ex as Exception, 'Error while closing consumer.');
      }
    }
  }

  ///  Clear method are clears component state.
  ///  Parameters:
  ///  - [correlationId] (optional) transaction id to trace execution through call chain.
  /// Return            Future that recive a null when clean compleate
  /// Throws error
  @override
  Future clear(String? correlationId) async {
    _checkOpened('');

    var count = 0;
    if (_queue != null) {
      count = _queue!.messageCount;
      await _queue!.purge();
    }

    logger.trace(
        correlationId, 'Cleared  %s messages in _queue %s', [count, name]);
  }
}
